Applications/Utils/MeshGeoTools/ComputeSurfaceNodeIDsInPolygonalRegion.cpp

```diff
@@ -0,0 +1,161 @@
+ *              http://www.opengeosys.org/LICENSE.txt
+ */
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "tclap/CmdLine.h"
+#include "logog/include/logog.hpp"
+
+#include "Applications/ApplicationsLib/LogogSetup.h"
+
+#include "BaseLib/LogogSimpleFormatter.h"
+#include "BaseLib/StringTools.h"
+#include "BaseLib/FileTools.h"
+
+#include "FileIO/readMeshFromFile.h"
+#include "FileIO/readGeometryFromFile.h"
+
+#include "GeoLib/GEOObjects.h"
+#include "GeoLib/Polygon.h"
+
+#include "MathLib/Vector3.h"
+
+#include "MeshLib/Mesh.h"
+#include "MeshLib/Node.h"
+#include "MeshLib/MeshSurfaceExtraction.h"
+
+void writeToFile(std::string const& id_area_fname, std::string const& csv_fname,
+	std::vector<std::pair<std::size_t, double>> const& ids_and_areas,
+	std::vector<MeshLib::Node*> const& mesh_nodes)
+{
+	std::ofstream ids_and_area_out(id_area_fname);
+	if (!ids_and_area_out) {
+		ERR("Unable to open the file \"%s\" - aborting.", id_area_fname.c_str());
+		std::abort();
+	}
+	std::ofstream csv_out(csv_fname);
+	if (!csv_out) {
+		ERR("Unable to open the file \"%s\" - aborting.", csv_fname.c_str());
+		std::abort();
+	}
+
+	ids_and_area_out << std::setprecision(20);
+	csv_out << std::setprecision(20);
+
+	ids_and_area_out << ids_and_areas[0].first << " " << ids_and_areas[0].second;
+	csv_out << "ID x y z area node_id\n"; // CSV header
+	csv_out << 0 << " " << *mesh_nodes[ids_and_areas[0].first]
+			<< ids_and_areas[0].second << " " << ids_and_areas[0].first;
+	for (std::size_t k(0); k<ids_and_areas.size(); k++) {
+		ids_and_area_out << "\n"
+			<< ids_and_areas[k].first << " " << ids_and_areas[k].second;
+		csv_out << "\n" << k << " " << *mesh_nodes[ids_and_areas[k].first]
+			<< ids_and_areas[k].second << " " << ids_and_areas[k].first;
+	}
+}
+
+int main (int argc, char* argv[])
+{
+	ApplicationsLib::LogogSetup logog_setup;
+
+	TCLAP::CmdLine cmd("Computes ids of mesh nodes that are in polygonal "
+		"regions and resides on the top surface. The polygonal regions have to "
+		"be given in a gml- or gli-file. The found mesh nodes and the associated"
+		" area are written as txt and csv data.", ' ', "0.1");
+	TCLAP::ValueArg<std::string> mesh_in("m", "mesh-input-file",
+		"the name of the file containing the input mesh", true,
+		"", "file name of input mesh");
+	cmd.add(mesh_in);
+	TCLAP::ValueArg<std::string> geo_in("g", "geo-file",
+		"the name of the gml file containing the polygons", true,
+		"", "file name of input geometry");
+	cmd.add(geo_in);
+
+	cmd.parse(argc, argv);
+
+	std::unique_ptr<MeshLib::Mesh const> mesh(FileIO::readMeshFromFile(mesh_in.getValue()));
+	INFO("Mesh read: %d nodes, %d elements.", mesh->getNNodes(), mesh->getNElements());
+
+	GeoLib::GEOObjects geo_objs;
+	FileIO::readGeometryFromFile(geo_in.getValue(), geo_objs);
+	std::vector<std::string> geo_names;
+	geo_objs.getGeometryNames(geo_names);
+	INFO("Geometry \"%s\" read: %d points, %d polylines.",
```

Comment:
Maybe wrong format character: "%d" is for signed integer, size_t will be unsigned long in many cases. Better use "%lu"? Do you get scrambled output with your format?